### PR TITLE
Mc 4759

### PR DESCRIPTION
**Pull request by kingmushroom**
Should fix 4759 and 4753.

Note though, the strata updates would need to have taken place first before this could merge in.

### CODE REVIEW

**Comment by kingmushroom** [comment]
yep, my bad, forgot to change that bit

File: calculate_means_wrangler.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 from es_aws_functions import aws_functions, exception_classes
 from marshmallow import Schema, fields
 
+import imputation_functions as impfunc
```
Comment (KrisRogos) [inline calculate_means_wrangler.py]: In atypicals and calculate movement this is imported as `imp_func`, please make consistent

File: dev-requirements.txt [hunk 2]
```diff
@@ -105,4 +105,6 @@ werkzeug==0.15.4
 wrapt==1.11.1
 xmltodict==0.12.0
 zipp==0.5.1
-git+https://github.com/ONSdigital/es-aws-functions.git
\ No newline at end of file
+git+https://github.com/ONSdigital/es-aws-functions.git
+
+# Test Sausage
```
Comment (KrisRogos) [inline dev-requirements.txt]: 🌭

File: apply_factors_wrangler.py [hunk 3]
```diff
@@ -183,7 +191,7 @@ def lambda_handler(event, context):
             )
         )
 
-        filtered_data = final_imputed.drop(cols_to_drop, axis=1)
+        filtered_data = final_imputed.drop(cols_to_drop, axis=1, errors='ignore')
```
Comment (Piwington) [inline apply_factors_wrangler.py]: Due to our changes to Calculate Factors, the only columns that need dropping are the "imputation_factor_" columns. As such can we refactor the cols_to_drop variable to reflect this and we would no longer need to use, errors='ignore'.

File: apply_factors_wrangler.py [hunk 4]
```diff
@@ -108,15 +110,20 @@ def lambda_handler(event, context):
             on="responder_id",
         )
         logger.info("Successfully merged previous period data with non-responder df")
+
+        # filter factors df to only get current period
+        factors_dataframe = factors_dataframe[
+            factors_dataframe['period'] == int(current_period)]
```
Comment (Piwington) [inline apply_factors_wrangler.py]: Shouldn't have a hardcoded period_column name and should use the RuntimeVariable instead.

File: calculate_imputation_factors_wrangler.py [hunk 5]
```diff
@@ -91,8 +97,15 @@ def lambda_handler(event, context):
         if not json_response['success']:
             raise exception_classes.MethodFailure(json_response['error'])
 
+        output_df = pd.read_json(json_response['data'])
+        distinct_values.append(period_column)
```
Comment (Piwington) [inline calculate_imputation_factors_wrangler.py]: Related to the above comment. Do we even need period now as this is a dataframe of factors, region and strata?

File: tests/test_apply_factors.py [hunk 6]
```diff
@@ -273,6 +279,15 @@ def test_method(self):
...
                 "json_data": json.loads(methodinput.to_json(orient="records")),
+                # Note, sum columns here actually overwrites 608 total
+                # but this way it checks both paths in sum_data_columns
+                "sum_columns": [{"column_name": "Q608_total", "data": {
+                                    "Q603_concreting_sand": "+",
+                                    "Q602_building_soft_sand": "+"}},
+                                {"column_name": "Q608_total", "data": {
+                                    "Q603_concreting_sand": "+",
+                                    "Q602_building_soft_sand": "-"}}
+                                ],
```
Comment (Piwington) [inline tests/test_apply_factors.py]: If we use this to test both paths we should do different columns (Or can specify a new column I believe). Then add a second assert to test the output.

Comment (kingmushroom) [reply tests/test_apply_factors.py]: This tests the paths. i added an extra set of variables to your test

File: apply_factors_wrangler.py (hunk 3 above)
Comment (kingmushroom) [reply apply_factors_wrangler.py]: quite right, gonna go test this on aws

**Comment by Piwington** [comment]
Added The Updated Drop Mike Tested To The Branch. Leaving The Other Change To Mike As Unsure Where He Got To With It.

**Comment by kingmushroom** [comment]
On conversation with david, we've decided to leave the period filter in apply factors as a belt & braces. If current and previous period is on the data at that point, it'll be filtered anyway.

Additionally, i noticed that movement method was still using env vars to pass in current n previous period. I've altered to pick up the runtimevariable and pass it to the method. There is now a previous period calculation function in movementmethod.

**Comment by kingmushroom** [comment]
#49 #48 #47

**Comment by kingmushroom** [comment]
#46

**Pull request merged**